mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java

```diff
@@ -208,28 +208,34 @@ public void delete(MessageId messageId) {
 
     @Override
     public Map<MailboxId, UpdatedFlags> setFlags(MessageId messageId, List<MailboxId> mailboxIds, Flags newState, MessageManager.FlagsUpdateMode updateMode) throws MailboxException {
-        CassandraMessageId cassandraMessageId = (CassandraMessageId) messageId;
-
         return mailboxIds.stream()
-            .map(mailboxId -> (CassandraId) mailboxId)
-            .flatMap(mailboxId -> imapUidDAO.retrieve(cassandraMessageId, Optional.of(mailboxId)).join())
-            .map(composedMessageId -> flagsUpdateWithRetry(newState, updateMode, composedMessageId))
+            .distinct()
+            .filter(mailboxId -> imapUidDAO.retrieve((CassandraMessageId) messageId, Optional.of((CassandraId) mailboxId))
+                .join()
+                .findAny()
+                .isPresent())
+            .map(mailboxId -> flagsUpdateWithRetry(newState, updateMode, mailboxId, messageId))
+            .filter(Optional::isPresent)
```

Comment:
flagsUpdateWithRetry being a private method, it could return a Stream on zero or one element and flatmap here